File: 015Triangle_minimum_path_sum/015Triangle_minimum_path_sum.py
class Solution:
    def minimumTotal(self, triangle: [[int]]) -> int:
        n = len(triangle)
        f = list()
        f.append(triangle[0][0])
        for i in range(1, n):
            f.append(f[i - 1] + triangle[i][i])
            for j in range(i - 1, 0, -1):
                f[j] = min(f[j - 1], f[j]) + triangle[i][j]
            f[0] = f[0] + triangle[i][0]
        return min(f)


triangle = [[2], [3, 4], [6, 5, 7], [4, 1, 8, 3]]
func = Solution()
print(func.minimumTotal(triangle))

File: 015Triangle_minimum_path_sum/015Triangle_minimum_path_sum.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using std::cout;
using std::endl;
using std::min;
using std::vector;

class Solution
{
public:
    int minimumTotal(vector<vector<int>> &triangle)
    {
        int n = triangle.size();
        vector<int> f(n);
        f[0] = triangle[0][0];
        for (int i = 1; i < n; ++i)
        {
            // ??????????????????ß“???
            f[i] = f[i - 1] + triangle[i][i];
            // ??????ß“???
            for (int j = i - 1; j > 0; --j)
            {
                f[j] = min(f[j - 1], f[j]) + triangle[i][j];
            }
            // ??ß÷?????????????
            f[0] += triangle[i][0];
        }
        // ??¡Ÿ????n??n??????ß÷????
        return *min_element(f.begin(), f.end());
    }
};

int main()
{
    vector<vector<int>> triangle = {
        {2},
        {3, 4},
        {6, 5, 7},
        {4, 1, 8, 3}};
    Solution func;
    cout << "the minimum path sum:" << func.minimumTotal(triangle) << endl;
    return 0;
}